sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbGridFSIO.java

```diff
@@ -0,0 +1,386 @@
+    }
+
+    @Override
+    public PCollection<T> apply(PBegin input) {
+      org.apache.beam.sdk.io.Read.Bounded<T> unbounded =
+          org.apache.beam.sdk.io.Read.from(options);
+      PCollection<T> output = input.getPipeline().apply(unbounded);
+      if (options.coder != null) {
+        output.setCoder(options.coder);
+      }
+      return output;
+    }
+
+    static class BoundedGridFSSource<T> extends BoundedSource<T> {
+      private final String uri;
+      private final String database;
+      private final String bucket;
+      private final ParseCallback<T> parser;
+      private final Coder<T> coder;
+
+      @Nullable
+      private List<ObjectId> objectIds;
+
+      private transient Mongo mongo;
+      private transient GridFS gridfs;
+
+
+      BoundedGridFSSource(String uri, String database, String bucket,
+          ParseCallback<T> parser, Coder<T> coder) {
+        this.uri = uri;
+        this.database = database;
+        this.bucket = bucket;
+        this.parser = parser;
+        this.coder = coder;
+      }
+      BoundedGridFSSource(String uri, String database, String bucket,
+          ParseCallback<T> parser, Coder<T> coder, List<ObjectId> objectIds) {
+        this.uri = uri;
+        this.database = database;
+        this.bucket = bucket;
+        this.parser = parser;
+        this.coder = coder;
+        this.objectIds = objectIds;
+      }
+      private synchronized void setupGridFS() {
+        if (gridfs == null) {
+          mongo = uri == null ? new Mongo() : new Mongo(new MongoURI(uri));
+          DB db = database == null ? mongo.getDB("gridfs") : mongo.getDB(database);
+          gridfs = bucket == null ? new GridFS(db) : new GridFS(db, bucket);
+        }
+      }
+      private synchronized void closeGridFS() {
+        if (gridfs != null) {
+          gridfs = null;
+          mongo.close();
+          mongo = null;
+        }
+      }
+
+      @Override
+      public List<? extends BoundedSource<T>> splitIntoBundles(long desiredBundleSizeBytes,
+          PipelineOptions options) throws Exception {
+        try {
+          setupGridFS();
+          DBCursor cursor = gridfs.getFileList().sort(null);
+          long size = 0;
+          List<BoundedGridFSSource<T>> list = new LinkedList<>();
+          List<ObjectId> objects = new LinkedList<>();
+          while (cursor.hasNext()) {
+            GridFSDBFile file = (GridFSDBFile) cursor.next();
+            long len = file.getLength();
+            if ((size + len) > desiredBundleSizeBytes && !objects.isEmpty()) {
+              list.add(new BoundedGridFSSource<T>(uri, database, bucket, parser, coder, objects));
+              size = 0;
+              objects = new LinkedList<>();
+            }
+            objects.add((ObjectId) file.getId());
+            size += len;
+          }
+          if (!objects.isEmpty()) {
+            list.add(new BoundedGridFSSource<T>(uri, database, bucket, parser, coder, objects));
+          }
```

Comment:
There isn't usable access to the chunk collection so you cannot split on it easily.   The only supported access to the data is via the InputStream on the GridFSDBFile which is not random IO to allow a good split from within the file.  I also just added a query filter for the file list and that would also not be usable with the chunk collection as the chunks don't know the filenames they are associated with.